packages/gatsby/src/bootstrap/requires-writer.ts

```diff
@@ -176,8 +183,22 @@ export const writeAll = async (state: IGatsbyState): Promise<boolean> => {
   const pages = [...state.pages.values()]
   const matchPaths = getMatchPaths(pages)
   const components = getComponents(pages)
+  const clientVisitedPageComponents = [...state.clientVisitedPages.values()]
+
+  // Remove any page components that no longer exist.
+  const cleanedClientVisitedPageComponents = components.filter(c =>
+    clientVisitedPageComponents.some(s => s === c.componentChunkName)
+  )
```

Comment:
Scaling red flag. What are `n` and `m` here? What's the worst case runtime? It's fine if it's tiny but if it scales with the number of pages/nodes/content creating a `Set` first might be worth it.

Reply:
This is looping over page components which grows slowly even for very large sites (since they'll have most of their pages getting built by one page component). It's not the number of pages but the number of _page types_. I tested this a lot with a 1000 unique page components & didn't notice any slow downs.